modules/bezug_e3dc/e3dc.py

```diff
@@ -1,90 +1,37 @@
 #!/usr/bin/python
-import sys
-# import os
-# import time
-# import getopt
-# import socket
-# import ConfigParser
-import struct
-# import binascii
-from pymodbus.client.sync import ModbusTcpClient
+from pymodbus.constants import Endian
+import logging
+import math
+from typing import List
 
-ipaddress = str(sys.argv[1])
+from helpermodules.cli import run_using_positional_cli_args
+from helpermodules.log import setup_logging_stdout
+from modules.common.component_state import CounterState
+from modules.common.modbus import ModbusClient, ModbusDataType
+from modules.common.store import get_counter_value_store
 
-client = ModbusTcpClient(ipaddress, port=502)
+log = logging.getLogger("E3DC EVU")
 
-# evu punkt
-resp= client.read_holding_registers(40073,2,unit=1)
-value1 = resp.registers[0]
-value2 = resp.registers[1]
-all = format(value2, '04x') + format(value1, '04x')
-final = int(struct.unpack('>i', all.decode('hex'))[0])
-f = open('/var/www/html/openWB/ramdisk/wattbezug', 'w')
-f.write(str(final))
-f.close()
+def update(ipaddress: str):
+    log.debug("Beginning update")
+    client = ModbusClient(ipaddress, port=502)
+#40074 EVU Punkt negativ -> Einspeisung in Watt
+    power_all = client.read_holding_registers(40073, ModbusDataType.INT_32,wordorder=Endian.Little, unit=1)
+#40130 Phasenleistung in Watt
+    powers = [val for val in client.read_holding_registers(40129, [ModbusDataType.INT_16] * 3,unit=1)]
+    currents = [0]*3
+    voltages = [230]*3
```

Comment:
Wenn `voltages` nicht bekannt ist, dann komplett weglassen. Der Parameter für den `CounterState` ist optional. Dann wird auch der vorgesehene Wert für "unbekannt" (aktuell [0, 0, 0]) verwendet.

Reply:
@LKuemmel hat in der Klasse `CounterState` den Wert `[0, 0, 0]` als Wert vorgegeben, der verwendet wird, wenn der Wert unbekannt ist. Warum weiß ich nicht. Ich kann nur Vermutungen anstellen:

Ein Problem welches sich ergeben könnte ist, dass oWB andernfalls nicht unterscheiden kann, ob der Wert geraten ist oder tatsächlich ein Messwert ist. Das mag ggf. schon ein Unterschied sein. Wenn es zum Beispiel keinen Messwert vom EVU gibt, könnte man ersatzweise die Spannung vom Ladepunkt nehmen welche (wenn es eine oWB oder eine andere Hardware ist, die die Spannung hergibt) dann zur Verfügung steht. Wenn du aber hier 230V fest reinschreibst, dann hat die Software keine Ahnung, dass sie ersatzweise einen anderen Wert nehmen soll.

Ein weiterer Punkt ist, dass es vorteilhaft sein kann, wenn alle Module einheitlich arbeiten. Wenn E3DC hier 230V reinschreibt, aber ein anderes Modul nicht, dann kann das ganze für den Nutzer und auch für den Support ("ach so, nein, das wusste ich nicht, dass das E3DC Modul sich hier anders verhält als alle anderen Module") ein Stolperstein sein. Hier ist gerade die Idee davon, dass man einzelne Parameter für den `CounterState` weglassen kann, es einen global einheitlichen Wert für "unbekannt" gibt.

Tatsächlich frage ich mich auch, ob man hier die errechneten `currents` angeben sollte, oder ob nicht besser wäre diese Werte bei nicht vorhandensein an anderer Stelle in der Software zu berechnen. (zum Beispiel in die Klasse `CounterState` oder alternativ immer dann, wenn gerade irgendwo ein Wert gebraucht wird).

Wenn man sich eine Weltkarte ansieht, welche Spannung so verwendet wird:
![Weltkarte Spannung&Frequenz](https://upload.wikimedia.org/wikipedia/commons/7/70/World_Map_of_Mains_Voltages_and_Frequencies%2C_Detailed.svg)
Dann fällt auf, dass bis auf Monaco und den Färöer in Europa einheitlich bei 230V@50Hz angesagt ist. Die Software von openWB ist bis jetzt nur auf Deutsch und es erweckt nicht den Eindruck, als hätte das oWB-Team auch nur ansatzweise im Sinn für andere Kunden als Deutsche interessant zu sein und schon garnicht Europa zu verlassen. Dennoch ist dann die Frage: Will man sich den Weg dahin verbauchen in dem man die Netzspannung in Modulen hardkodiert? Es ist ja nicht schwierig sich den Weg offen zu halten. Man nimmt dann eben eine Konstante, die in diesem Fall als Default-Wert im `CounterState` hinterlegt ist und bei Bedarf kann man diese Konstante an zentraler Stelle leicht abwandeln.

Wenn du willst, dass die Spannung mit angezeigt wird, dann wäre es in meinen Augen sinnvoller die Klasse `CounterState` zu ändern und hier im Modul einen Alleingang zu machen.

Das ist aber nur meine Meinung. Entscheiden muss darüber Lena. Oder Lutz.

Reply:
Der Standardwert für die Spannung beträgt nun 230V. Fehlende Ströme werden seit PR #1921 automatisch berechnet.
@okaegi  Kannst Du die Berechnung bitte noch entfernen? Dann kann der PR gemergt werden.

Reply:
mach ich über das Wochenende,
Gruss Oliver

Reply:
Berechnung für Current pro Phase und setting Volt pro Phase entfernt, kann nun übernommen werden